src/Documenter.jl

```diff
@@ -555,11 +555,23 @@ function git_push(
     end
 
     if authentication_method(deploy_config) === SSH
-        # Extract host from repo as everything up to first ':' or '/' character
-        host = match(r"(.*?)[:\/]", repo)[1]
-
-        # The upstream URL to which we push new content and the ssh decryption commands.
-        upstream = "git@$(replace(repo, "$host/" => "$host:"))"
+        # Create the upstream repository
+        if '@' in repo
+            # For repo URL of the form user@host:/path/to/repo, extract user
+            user, rest = split(repo, '@')
+            # Extract host from repo as everything between @ and first ':' or '/' character
+            host = match(r"(.*?)[:\/]", rest)[1]
+            # The upstream URL to which we push new content and the ssh decryption commands.
+            upstream = "$(replace(repo, "$host/" => "$host:"))"
+        else
+            # if no user given, default to "git"
+            user = "git"
+            # Extract host from repo as everything up to first ':' or '/' character
+            host = match(r"(.*?)[:\/]", repo)[1]
+            # The upstream URL to which we push new content and the ssh decryption commands.
+            upstream = "git@$(replace(repo, "$host/" => "$host:"))"
+        end
```

Comment:
Just a suggestion: it feels like it should be possible to do this with just a single regex, e.g.:

```julia
match(r"(?:([^@]*)@)?(.*?)[:\/]", "user@foo.com/blah.git")
```

And if the first capture group is `nothing`, then use the default username.

Reply:
Also, could we refactor this into a function (returning a `user, upstream` tuple?) and add some unit tests, just to make sure that the code does exactly what it is supposed to?

Reply:
Yes, that is much more elegant (my feeble regex skills failed me). Where would be a good place to put the unit tests?

Reply:
We can add it here: https://github.com/JuliaDocs/Documenter.jl/blob/master/docs/src/lib/internals/documenter.md